### ISSUE TITLE
Add option to make to_dict of an element that may contain attributes always return a dict

### ISSUE THREAD

**Issue by pythonator** [brunato/xmlschema#108]
The library currently returns the string `text` for
```python
schema.to_dict("""<root>text</root>""")
```
and returns the dictionary `{'@attr': 'attr_value', '$': 'text'}` for
```python
schema.to_dict("""<root attr="attr_value">text</root>""")
```
given a schema where an element may or may not have an attribute
```xml
<?xml version="1.0" encoding="utf-8"?>
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">
  <xs:element name="root" type="root"/>
  <xs:complexType name="root">
    <xs:simpleContent>
      <xs:extension base="xs:string">
        <xs:attribute name="attr" type="xs:string"/>
        <xs:attribute name="attrWithDefault" type="xs:string" default="default_value"/>
        <xs:attribute name="attrWithFixed" type="xs:string" fixed="fixed_value"/>
      </xs:extension>
    </xs:simpleContent>
  </xs:complexType>
</xs:schema>
```

This makes it hard to use the return value of `to_dict` since we need to test whether it is a dictionary or a string/int/etc. Furthermore, this ignores the `fixed` and `default` values of the attribute.

I think the library should have the option to force such elements to be converted into a dictionary. Perhaps this should be the default behavior?

If an attribute is missing and the schema has a `default` or `fixed`, then that value should be used. Otherwise it should use `None`. In the previous example

```python
schema.to_dict("""<root>text</root>""")
```

would be decoded as
```python
{'@attr': None,
 '@attrWithDefault': 'default_value',
 '@attrWithFixed': 'fixed_value',
 '$': 'text'}
```

**Comment by brunato** [comment]
Hi,
it's difficult to satisfy all the different needs on XML decoding into a data structure (or JSON) with only one convention, so the default converter class can be changed at schema building or as decoding option. In the library there are some different converter classes implemented in module *converter.py*.

Had you already seen these alternative converters and the wiki page [JSON and XML Conversion](http://wiki.open311.org/JSON_and_XML_Conversion/)?

**Comment by pythonator** [comment]
Yes I have seen the other converters and the wiki page. None of the converters give the behavior I described.

Since this library has all the information in the XSD, it is smarter than simply converting xml to json using one of the conversions in the wiki page.
For example, an element with `maxOccurs="2"` will be converted to a list, even if there is only one tag in the xml document.

What I am requesting is similar. Even if an attribute doesn't appear in the XML, we know that it can, and we know its default/fixed value, so we could add it to the JSON document.

**Comment by brunato** [comment]
Converters were projected to include all the options about the form of the data structure produced by the XML decoding, avoiding the adding of options like *force_list* or similar. This because the derived classes, that implement well defined conventions, define their formats not in base of the schema but in respect of the XML input data. 

So in the base class I kept only options that can alter a name or a value or the container type, not the structure form. Instead of changing the default behaviour, that for example currently implies that an empty simple content element is mapped to `None` (also in this case the converter should returns a dictionary like in the case of a text without attributes?), I think it's better to create a schema-strict converter class that behaves like you described. I would call this *StrictConverter* (other proposals are welcome :-) ). What do you think about this?

The missing of default/fixed values is a bug that I have to fix. Infact the defaults are already inserted in the case of a simple type element, but not in a case of a simple content element.

**Comment by brunato** [comment]
Instead of defining a new converter type is better to add a more general options on decoding, for filling missing attributes and decoding errors. These option will be called *filler* (in case you want to replace a missing/wrong value with a value of a specific type, instead of `None`) and *fill_missing* for fill also missing attributes.

**Comment by pythonator** [comment]
So you're saying not to implement *StrictConverter* and instead add more general options to `XMLSchemaConverter`, even if they may change the dict format based on the schema in addition to the XML input data?

I am not sure I understand what you meant by the *filler* option. There may not be a filler value that will pass all the validations for all the schema types.

Here are a few possible schema-dependent options that I thought about:

- *fill_defaults* and *fill_fixed* - These give missing attributes the value written in the schema's default/fixed. I think these options should be merged into a single option named *fill_defaults*.

To solve the problem where the output is sometimes a string and sometimes a dictionary I would add one of the following options:
- *fill_missing_attrs* - This adds all missing attributes to the output with the value None.
- *force_dicts* - This would force types which may have attributes to always be formatted as dictionaries. Missing attributes would not appear in the dictionary.

You also mentioned the *force_list* option, I also think that it could be useful.

**Comment by brunato** [comment]
The *filler* optional argument will be a callable that takes an argument (XSD simple content element or an XSD attribute). When *filler* is not provided the decoder uses the value `None` (like now, that remain the default behaviour). 

The concept of missing apply clearly only to attributes, there is not a general and deterministic rule for deciding when an element is missing. I prefer *fill_missing* instead of *fill_missing_attrs* for brevity and because is a more general concept that can be eventually extended in the future.
The option *use_defaults* already cover the cases when an attribute value is missing or an element value is `None`. Consider that these are decoder's options that have to be propagated down with kwargs and examined at many levels.

About `StrictConverter`: I've tried to write it down but something is not clear to me about the exact format to produce. In my develop I've already extended the base converter class with *preserve_root*, previously  related only to the class *ParkerConverter*. Considering that adding options to the base class is not a drama (the converter instance is create once at the start of the decoding process), I will examine the feasability of adding a `force_dict=False` (for all the complex types?) and `force_list=False` to the `XMLSchemaConverter`. Those options will be used probably only by the default converter but could be useful for new derived converter types.
